lib/codecs/src/decoding/format/protobuf.rs

```diff
@@ -0,0 +1,311 @@
+            .map_err(|error| format!("Error parsing protobuf: {:?}", error))?;
+
+        let proto_vrl = to_vrl(
+            prost_reflect::Value::Message(dynamic_message),
+            &prost_reflect::Kind::Message(self.message_descriptor.clone()),
+        )?;
+        let mut event = Event::Log(LogEvent::from(proto_vrl));
+        let event = match log_namespace {
+            LogNamespace::Vector => event,
+            LogNamespace::Legacy => {
+                let timestamp = Utc::now();
+
+                if let Some(timestamp_key) = log_schema().timestamp_key() {
+                    let log = event.as_mut_log();
+                    if !log.contains((PathPrefix::Event, timestamp_key)) {
+                        log.insert((PathPrefix::Event, timestamp_key), timestamp);
+                    }
+                }
+
+                event
+            }
+        };
+
+        Ok(smallvec![event])
+    }
+}
+
+impl From<&ProtobufDeserializerConfig> for ProtobufDeserializer {
+    fn from(config: &ProtobufDeserializerConfig) -> Self {
+        Self::new(config.desc_file.clone(), config.message_type.clone())
+    }
+}
+
+fn to_vrl(
+    prost_reflect_value: prost_reflect::Value,
+    kind: &prost_reflect::Kind,
+) -> vector_common::Result<vrl::value::Value> {
+    let vrl_value = match prost_reflect_value {
+        prost_reflect::Value::Bool(v) => vrl::value::Value::from(v),
+        prost_reflect::Value::I32(v) => vrl::value::Value::from(v),
+        prost_reflect::Value::I64(v) => vrl::value::Value::from(v),
+        prost_reflect::Value::U32(v) => vrl::value::Value::from(v),
+        prost_reflect::Value::U64(v) => vrl::value::Value::from(v),
+        prost_reflect::Value::F32(v) => vrl::value::Value::Float(
+            NotNan::new(f64::from(v)).map_err(|_e| format!("Float number cannot be Nan"))?,
+        ),
+        prost_reflect::Value::F64(v) => vrl::value::Value::Float(
+            NotNan::new(v).map_err(|_e| format!("F64 number cannot be Nan"))?,
+        ),
+        prost_reflect::Value::String(v) => vrl::value::Value::from(v),
+        prost_reflect::Value::Bytes(v) => vrl::value::Value::from(v),
+        prost_reflect::Value::EnumNumber(v) => {
+            let enum_desc = kind.as_enum().ok_or_else(|| format!("Internal error while parsing protobuf enum"))?;
+            vrl::value::Value::from(
+                enum_desc
+                    .get_value(v)
+                    .ok_or_else(|| format!("The number {} cannot be in '{}'", v, enum_desc.name()))?
+                    .name(),
+            )
+        }
+        prost_reflect::Value::Message(mut v) => {
+            let mut obj_map = BTreeMap::new();
+            for field_desc in v.descriptor().fields() {
+                let field = v.get_field_mut(&field_desc);
+                let mut taken_value = prost_reflect::Value::Bool(false);
+                std::mem::swap(&mut taken_value, field);
```

Comment:
Did you try [this](https://docs.rs/prost-reflect/0.11.4/prost_reflect/struct.DynamicMessage.html#method.get_field)? It returns a `Cow`. See suggestion below. We can also get rid of `mut` this way.

Reply:
that was the original call, but I wanted to avoid unneeded copy and allocations, I tried to "steal" all the resources from the value itself.
you can see the difference in `to_vrl_by_ref` in my `cargo bench` demo.